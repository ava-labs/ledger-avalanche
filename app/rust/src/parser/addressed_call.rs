use crate::parser::{AddressedCallPayload, FromBytes, ParserError};
use core::mem::MaybeUninit;
use nom::{
    bytes::complete::take,
    number::complete::{be_u16, be_u32},
};

#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub struct AddressedCall<'b> {
    pub source_address_size: u32,
    pub source_address: &'b [u8],
    pub message_type: u32,
    pub payload: AddressedCallPayload<'b>,
}

impl<'b> FromBytes<'b> for AddressedCall<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("AddressedCall::from_bytes_into\x00");

        let out = out.as_mut_ptr();

        // codec_id (2 bytes)
        let (rem, codec_id) = be_u16(input)?;
        if codec_id != 0 {
            return Err(nom::Err::Error(ParserError::InvalidCodecId));
        }

        // type_id (4 bytes)
        let (rem, type_id) = be_u32(rem)?;
        if type_id != 1 {
            return Err(nom::Err::Error(ParserError::InvalidTypeId));
        }

        // source address size (4 bytes)
        let (rem, source_address_size) = be_u32(rem)?;

        // source address (variable size)
        let (rem, source_address) = take(source_address_size as usize)(rem)?;

        // payload_size (4 bytes)
        let (rem, _) = be_u32(rem)?;

        // payload (remaining bytes) - parse as AddressedCallPayload
        let payload = AddressedCallPayload::from_payload(rem)?;

        unsafe {
            core::ptr::addr_of_mut!((*out).source_address_size).write(source_address_size);
            core::ptr::addr_of_mut!((*out).source_address).write(source_address);
            core::ptr::addr_of_mut!((*out).payload).write(payload);
        }

        Ok(&[]) // All bytes consumed
    }
}
