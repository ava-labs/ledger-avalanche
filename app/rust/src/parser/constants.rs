/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/

//network ids
pub const NETWORK_ID_MAINNET: u32 = 1;
pub const NETWORK_ID_FUJI: u32 = 5;
pub const NETWORK_ID_LOCAL: u32 = 12345;
pub const NETWORK_ID_CUSTOM: u32 = 1337;
pub const NETWORK_ID_DEVNET: u32 = 76;
pub const ETH_MAINNET_ID: u64 = 1;

// hrp
pub const HRP_MAINNET: &str = "avax";
pub const HRP_TESTNET: &str = "fuji";
pub const HRP_LOCAL: &str = "local";
pub const HRP_CUSTOM: &str = "custom";
pub const HRP_DEVNET: &str = "custom";
pub const TRANSFER_TX: u32 = 0;

// pvm transaction types
pub const PVM_CREATE_SUBNET: u32 = 0x00000010;
pub const PVM_EXPORT_TX: u32 = 0x00000012;
pub const PVM_IMPORT_TX: u32 = 0x00000011;
pub const PVM_ADD_VALIDATOR: u32 = 0x0000000c;
pub const PVM_ADD_SUBNET_VALIDATOR: u32 = 0x0000000d;
pub const PVM_ADD_DELEGATOR: u32 = 0x0000000e;
pub const PVM_CREATE_CHAIN: u32 = 0x0000000f;
pub const PVM_REMOVE_SUBNET_VALIDATOR: u32 = 0x00000017;
pub const PVM_TRANSFORM_SUBNET: u32 = 0x00000018;
pub const PVM_ADD_PERMISSIONLESS_VALIDATOR: u32 = 0x00000019;
pub const PVM_ADD_PERMISSIONLESS_DELEGATOR: u32 = 0x0000001A;

pub const PVM_CONVERT_SUBNET_L1: u32 = 0x00000023;
pub const PVM_REGISTER_L1_VALIDATOR: u32 = 0x00000024;
pub const PVM_SET_L1_VALIDATOR_WEIGHT: u32 = 0x00000025;
pub const PVM_DISABLE_L1_VALIDATOR: u32 = 0x00000027;
pub const PVM_INCREASE_L1_VALIDATOR_BALANCE: u32 = 0x00000026;

//  base tx for p-chain which collides with evm import tx.
//  we should check chain-id
pub const PVM_BASE_TX: u32 = 0x00000000;
pub const PVM_BASE_TX_TRANSFER: u32 = 0x00000022;

// avm transaction types
pub const AVM_CREATE_ASSET_TX: u32 = 0x00000001;
pub const AVM_OPERATION_TX: u32 = 0x00000002;
pub const AVM_IMPORT_TX: u32 = 0x00000003;
pub const AVM_EXPORT_TX: u32 = 0x00000004;

// evm transaction types
pub const EVM_IMPORT_TX: u32 = 0x00000000;
pub const EVM_EXPORT_TX: u32 = 0x00000001;
pub const EIP1559_TX: u8 = 0x02;
pub const EIP2930_TX: u8 = 0x01;
pub const ETH_ARG_LEN: usize = 32;
// The number of bytes to be shown
// when rendering the contract data
// It is cropped as this data can be
// very large.
pub const DEPLOY_DATA_PREVIEW_LEN: usize = 30;
// taken from app-ethereum where its value is 70
// but we reduce it to 50 to save some bytes
pub const COLLECTION_NAME_MAX_LEN: usize = 50;

//Avax units
pub const NANO_AVAX_DECIMAL_DIGITS: usize = 9;
pub const WEI_NAVAX_DIGITS: usize = 9;
pub const WEI_AVAX_DIGITS: usize = 18;
pub const DELEGATION_FEE_DIGITS: usize = 4;

// data formatting constants

// the lenght required to format a date like:
// yyyy-mm-dd hh:mm:ss UTC
pub const FORMATTED_STR_DATE_LEN: usize = 23;

// other constants
pub const CB58_CHECKSUM_LEN: usize = 4;
pub const U32_SIZE: usize = std::mem::size_of::<u32>();
pub const U64_SIZE: usize = std::mem::size_of::<u64>();

// The maximun lenght of a message chunk
// to be displayed on the device. each chunk
// would contain a formatted part of the original message
// this applies only in cases where the received
// message is in ASCII format
pub const MSG_MAX_CHUNK_LEN: usize = 100;

// types
pub type OutputIdx = u64;
