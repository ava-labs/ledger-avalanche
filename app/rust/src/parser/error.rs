use arrayvec::CapacityError;
use nom::error::ErrorKind;

#[repr(u32)]
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
// #[cfg_attr(test, derive(Debug))]
pub enum ParserError {
    ParserOk = 0,
    // Generic errors
    NoData,
    DisplayIdxOutOfRange,
    DisplayPageOutOfRange,
    UnexpectedError,
    ParserInitContextEmpty,
    ParserContextMismatch,
    // Required fields
    // Coin specific
    InvalidHashMode,
    InvalidSignature,
    InvalidPubkeyEncoding,
    InvalidAddressVersion,
    InvalidAddressLength,
    InvalidTypeId,
    InvalidCodec,
    InvalidThreshold,
    InvalidNetworkId,
    InvalidChainId,
    InvalidAsciiValue,
    InvalidTimestamp,
    InvalidStakingAmount,
    UnexpectedType,
    InvalidTransactionType,
    OperationOverflows,
    UnexpectedBufferEnd,
    UnexpectedNumberItems,
    UnexpectedField,
    ValueOutOfRange,
    InvalidAddress,
    InvalidPath,
    TooManyOutputs,
    InvalidAvaxMessage,
    UnexpectedData,
    InvalidEthMessage,
    InvalidEthSelector,
    InvalidAssetCall,
    NftInfoNotProvided,
    InvalidContractAddress,
    InvalidMessageSize,
    InvalidCodecId,
    InvalidLength,
    InvalidSourceAddressSize,
}

impl From<ErrorKind> for ParserError {
    fn from(err: ErrorKind) -> Self {
        match err {
            ErrorKind::Eof => ParserError::UnexpectedBufferEnd,
            ErrorKind::Permutation => ParserError::UnexpectedType,
            ErrorKind::TooLarge => ParserError::ValueOutOfRange,
            ErrorKind::Tag => ParserError::InvalidTypeId,
            _ => ParserError::UnexpectedError,
        }
    }
}

impl<I> nom::error::ParseError<I> for ParserError {
    fn from_error_kind(_input: I, kind: ErrorKind) -> Self {
        Self::from(kind)
    }

    // We don't have enough memory resources to use here an array with the last
    // N errors to be used as a backtrace, so that, we just propagate here the latest
    // reported error
    fn append(_input: I, _kind: ErrorKind, other: Self) -> Self {
        other
    }
}
impl From<ParserError> for nom::Err<ParserError> {
    fn from(error: ParserError) -> Self {
        nom::Err::Error(error)
    }
}

impl From<CapacityError> for ParserError {
    fn from(_error: CapacityError) -> Self {
        ParserError::UnexpectedBufferEnd
    }
}

impl From<nom::Err<Self>> for ParserError {
    fn from(e: nom::Err<Self>) -> Self {
        match e {
            nom::Err::Error(e) => e,
            nom::Err::Failure(e) => e,
            nom::Err::Incomplete(_) => Self::UnexpectedBufferEnd,
        }
    }
}
