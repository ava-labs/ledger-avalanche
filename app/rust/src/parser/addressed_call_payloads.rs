use crate::parser::{FromBytes, NodeId, ParserError, PchainOwner, SubnetId};
use core::mem::MaybeUninit;
use core::ptr::addr_of_mut;
use nom::{
    bytes::complete::take,
    number::complete::{be_u16, be_u32, be_u64},
};

// RegisterL1ValidatorMessage structure
#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub struct RegisterL1ValidatorMessage<'b> {
    pub subnet_id: SubnetId<'b>,
    pub node_id: NodeId<'b>,
    pub bls_pubkey: &'b [u8; 48],
    pub expiry: u64,
    pub remaining_balance_owner: PchainOwner<'b>,
    pub disable_owner: PchainOwner<'b>,
    pub weight: u64,
}

impl<'b> FromBytes<'b> for RegisterL1ValidatorMessage<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("RegisterL1ValidatorMessage::from_bytes_into\x00");

        let out = out.as_mut_ptr();

        // subnet_id
        let subnet_id = unsafe { &mut *core::ptr::addr_of_mut!((*out).subnet_id).cast() };
        let rem = SubnetId::from_bytes_into(input, subnet_id)?;

        // Node id size, not used
        let (rem, _node_size) = be_u32(rem)?;

        // node_id
        let node_id = unsafe { &mut *addr_of_mut!((*out).node_id).cast() };
        let rem = NodeId::from_bytes_into(rem, node_id)?;

        // bls_pubkey
        let (rem, bls_pubkey) = take(48usize)(rem)?;
        let bls_pubkey = arrayref::array_ref!(bls_pubkey, 0, 48);

        // expiration
        let (rem, expiry) = be_u64(rem)?;

        // remaining_balance_owner
        let remaining_balance_owner =
            unsafe { &mut *core::ptr::addr_of_mut!((*out).remaining_balance_owner).cast() };
        let rem = PchainOwner::from_bytes_into(rem, remaining_balance_owner)?;

        // disable_owner
        let disable_owner = unsafe { &mut *core::ptr::addr_of_mut!((*out).disable_owner).cast() };
        let rem = PchainOwner::from_bytes_into(rem, disable_owner)?;

        // weight
        let (rem, weight) = be_u64(rem)?;

        unsafe {
            core::ptr::addr_of_mut!((*out).bls_pubkey).write(bls_pubkey);
            core::ptr::addr_of_mut!((*out).expiry).write(expiry);
            core::ptr::addr_of_mut!((*out).weight).write(weight);
        }

        Ok(rem)
    }
}

// SetL1ValidatorWeightMessage structure
#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub struct SetL1ValidatorWeightMessage<'b> {
    pub codec_id: u16,
    pub type_id: u32,
    pub validation_id: &'b [u8; 32],
    pub nonce: u64,
    pub weight: u64,
}

impl<'b> FromBytes<'b> for SetL1ValidatorWeightMessage<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("SetL1ValidatorWeightMessage::from_bytes_into\x00");

        let out = out.as_mut_ptr();

        // validation_id
        let (rem, validation_id) = take(32usize)(input)?;
        let validation_id = arrayref::array_ref!(validation_id, 0, 32);

        // nonce
        let (rem, nonce) = be_u64(rem)?;

        // weight
        let (rem, weight) = be_u64(rem)?;

        unsafe {
            core::ptr::addr_of_mut!((*out).validation_id).write(validation_id);
            core::ptr::addr_of_mut!((*out).nonce).write(nonce);
            core::ptr::addr_of_mut!((*out).weight).write(weight);
        }

        Ok(rem)
    }
}

// Enum to represent different L1 validator message types
#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub enum AddressedCallPayload<'b> {
    RegisterL1Validator(RegisterL1ValidatorMessage<'b>),
    SetL1ValidatorWeight(SetL1ValidatorWeightMessage<'b>),
}

impl<'b> AddressedCallPayload<'b> {
    pub fn from_payload(payload: &'b [u8]) -> Result<Self, ParserError> {
        if payload.len() < 6 {
            return Err(ParserError::InvalidLength);
        }

        // Read codec_id
        let (rem, codec_id) = be_u16(payload)?;
        if codec_id != 0 {
            return Err(ParserError::InvalidCodecId);
        }

        // Read type_id
        let (rem, type_id) = be_u32(rem)?;

        match type_id {
            1 => {
                let mut msg = core::mem::MaybeUninit::uninit();
                let _ = RegisterL1ValidatorMessage::from_bytes_into(rem, &mut msg)?;
                let msg = unsafe { msg.assume_init() };
                Ok(AddressedCallPayload::RegisterL1Validator(msg))
            }
            3 => {
                let mut msg = core::mem::MaybeUninit::uninit();
                let _ = SetL1ValidatorWeightMessage::from_bytes_into(rem, &mut msg)?;
                let msg = unsafe { msg.assume_init() };
                Ok(AddressedCallPayload::SetL1ValidatorWeight(msg))
            }
            _ => Err(ParserError::InvalidTypeId),
        }
    }
}
