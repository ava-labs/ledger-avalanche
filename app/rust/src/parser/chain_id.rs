/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use crate::handlers::handle_ui_message;
use crate::sys::ViewError;
use core::{mem::MaybeUninit, ptr::addr_of_mut};
use nom::bytes::complete::take;

use crate::parser::{DisplayableItem, FromBytes, ParserError};
use crate::utils::hex_encode;

pub const CHAIN_ID_LEN: usize = 32;

#[derive(Clone, Copy, PartialEq, Eq)]
#[cfg_attr(test, derive(Debug))]
pub struct ChainId<'b>(&'b [u8; CHAIN_ID_LEN]);

impl<'b> ChainId<'b> {
    pub fn new(id: &'b [u8; CHAIN_ID_LEN]) -> Self {
        Self(id)
    }
}

impl<'b> FromBytes<'b> for ChainId<'b> {
    #[inline(never)]
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("ChainId::from_bytes_into\x00");

        let (rem, chain_id) = take(CHAIN_ID_LEN)(input)?;
        // This would not fail as previous line ensures we take
        // the right amount of bytes, also the alignemnt is correct
        let chain_id = arrayref::array_ref!(chain_id, 0, CHAIN_ID_LEN);

        //good ptr and no uninit reads
        let out = out.as_mut_ptr();
        unsafe {
            addr_of_mut!((*out).0).write(chain_id);
        }

        Ok(rem)
    }
}

impl<'b> DisplayableItem for ChainId<'b> {
    fn num_items(&self) -> Result<u8, ViewError> {
        Ok(1)
    }

    fn render_item(
        &self,
        item_n: u8,
        title: &mut [u8],
        message: &mut [u8],
        page: u8,
    ) -> Result<u8, ViewError> {
        use bolos::{pic_str, PIC};

        if item_n != 0 {
            return Err(ViewError::NoData);
        }
        let prefix = pic_str!(b"0x"!);
        let label = pic_str!(b"Chain ID");
        title[..label.len()].copy_from_slice(label);

        // prefix
        let mut out = [0; CHAIN_ID_LEN * 2 + 2];
        let mut sz = prefix.len();
        out[..prefix.len()].copy_from_slice(&prefix[..]);

        sz += hex_encode(self.0, &mut out[prefix.len()..]).map_err(|_| ViewError::Unknown)?;

        handle_ui_message(&out[..sz], message, page)
    }
}
