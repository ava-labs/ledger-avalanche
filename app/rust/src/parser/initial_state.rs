/*******************************************************************************
*   (c) 2018-2024 Zondax AG
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
********************************************************************************/
use core::{convert::TryFrom, mem::MaybeUninit, ptr::addr_of_mut};
use nom::number::complete::be_u32;

use crate::parser::{AvmOutput, FromBytes, ObjectList, ParserError};

#[derive(Clone, Copy, PartialEq, Eq)]
#[cfg_attr(test, derive(Debug))]
#[repr(C)]
pub enum FxId {
    SECP256KAsset,
    NftAsset,
}

impl TryFrom<u32> for FxId {
    type Error = ParserError;
    fn try_from(value: u32) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(FxId::SECP256KAsset),
            1 => Ok(FxId::NftAsset),
            _ => Err(ParserError::UnexpectedType),
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
#[cfg_attr(test, derive(Debug))]
#[repr(C)]
pub struct InitialState<'b> {
    id: FxId,
    outputs: ObjectList<'b, AvmOutput<'b>>,
}

impl<'b> FromBytes<'b> for InitialState<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("InitialState::from_bytes_into\x00");

        let output = out.as_mut_ptr();
        // get FxId
        let (rem, id) = be_u32(input)?;
        let id = FxId::try_from(id)?;

        // get outputs
        let outputs = unsafe { &mut *addr_of_mut!((*output).outputs).cast() };
        let rem = ObjectList::<AvmOutput>::new_into(rem, outputs)?;

        //good ptr and no uninit reads
        unsafe {
            addr_of_mut!((*output).id).write(id);
        }

        Ok(rem)
    }
}
