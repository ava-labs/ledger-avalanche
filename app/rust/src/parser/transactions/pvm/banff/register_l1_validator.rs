use crate::parser::{nano_avax_to_fp_str, ADDRESS_LEN};
use crate::utils::hex_encode;
use crate::{
    handlers::handle_ui_message,
    parser::{
        pchain_owner::PchainOwner, proof_of_possession::BLS_SIGNATURE_LEN, BaseTxFields,
        DisplayableItem, FromBytes, Header, NodeId, ParserError, PvmOutput, SubnetId,
        PVM_REGISTER_L1_VALIDATOR,
    },
};
use bolos::PIC;
use core::{mem::MaybeUninit, ptr::addr_of_mut};
use nom::{
    bytes::complete::{tag, take},
    number::complete::{be_u16, be_u32, be_u64},
};
use zemu_sys::ViewError;

#[derive(Clone, Copy, PartialEq, Eq)]
#[repr(C)]
#[cfg_attr(test, derive(Debug))]
pub struct RegisterL1ValidatorTx<'b> {
    pub tx_header: Header<'b>,
    pub base_tx: BaseTxFields<'b, PvmOutput<'b>>,
    pub balance: u64,
    pub signer: &'b [u8; BLS_SIGNATURE_LEN],
    // RegisterL1ValidatorMessage payload
    pub codec_id: u16,
    pub type_id: u32,
    pub subnet_id: SubnetId<'b>,
    pub node_id: NodeId<'b>,
    pub bls_pubkey: &'b [u8; 48],
    pub expiry: u64,
    pub remaining_balance_owner: PchainOwner<'b>,
    pub disable_owner: PchainOwner<'b>,
    pub weight: u64,
}

impl<'b> FromBytes<'b> for RegisterL1ValidatorTx<'b> {
    fn from_bytes_into(
        input: &'b [u8],
        out: &mut MaybeUninit<Self>,
    ) -> Result<&'b [u8], nom::Err<ParserError>> {
        crate::sys::zemu_log_stack("RegisterL1ValidatorTx::from_bytes_into\x00");

        let (rem, _) = tag(PVM_REGISTER_L1_VALIDATOR.to_be_bytes())(input)?;

        let out = out.as_mut_ptr();

        // tx header
        let tx_header = unsafe { &mut *addr_of_mut!((*out).tx_header).cast() };
        let rem = Header::from_bytes_into(rem, tx_header)?;

        // base_tx
        let base_tx = unsafe { &mut *addr_of_mut!((*out).base_tx).cast() };
        let rem = BaseTxFields::<PvmOutput>::from_bytes_into(rem, base_tx)?;

        // balance
        let (rem, balance) = be_u64(rem)?;

        // signer
        let (rem, signer) = take(BLS_SIGNATURE_LEN)(rem)?;
        let signer = arrayref::array_ref!(signer, 0, BLS_SIGNATURE_LEN);

        // codec_id
        let (rem, codec_id) = be_u16(rem)?;

        // type_id
        let (rem, type_id) = be_u32(rem)?;
        if type_id != 1 {
            return Err(nom::Err::Error(ParserError::InvalidTypeId));
        }

        // subnet_id
        let subnet_id = unsafe { &mut *addr_of_mut!((*out).subnet_id).cast() };
        let rem = SubnetId::from_bytes_into(rem, subnet_id)?;

        // Node id size, not used
        let (rem, _unused_value) = take(4usize)(rem)?;

        // node_id
        let node_id = unsafe { &mut *addr_of_mut!((*out).node_id).cast() };
        let rem = NodeId::from_bytes_into(rem, node_id)?;

        // bls_pubkey
        let (rem, bls_pubkey) = take(48usize)(rem)?;
        let bls_pubkey = arrayref::array_ref!(bls_pubkey, 0, 48);

        // expiration
        let (rem, expiry) = be_u64(rem)?;

        // remaining_balance_owner
        let remaining_balance_owner =
            unsafe { &mut *addr_of_mut!((*out).remaining_balance_owner).cast() };
        let rem = PchainOwner::from_bytes_into(rem, remaining_balance_owner)?;

        // disable_owner
        let disable_owner = unsafe { &mut *addr_of_mut!((*out).disable_owner).cast() };
        let rem = PchainOwner::from_bytes_into(rem, disable_owner)?;

        // weight
        let (rem, weight) = be_u64(rem)?;

        unsafe {
            addr_of_mut!((*out).codec_id).write(codec_id);
            addr_of_mut!((*out).weight).write(weight);
            addr_of_mut!((*out).balance).write(balance);
            addr_of_mut!((*out).bls_pubkey).write(bls_pubkey);
            addr_of_mut!((*out).expiry).write(expiry);
            addr_of_mut!((*out).signer).write(signer);
        }

        Ok(rem)
    }
}

impl<'b> RegisterL1ValidatorTx<'b> {
    // Info at https://github.com/ava-labs/avalanchejs/blob/master/src/utils/getBurnedAmountByTx.ts
    fn fee(&'b self) -> Result<u64, ParserError> {
        let sum_inputs = self.base_tx.sum_inputs_amount()?;

        let base_outputs = self.base_tx.sum_outputs_amount()?;

        let mut fee = sum_inputs
            .checked_sub(base_outputs)
            .ok_or(ParserError::OperationOverflows)?;

        fee = fee
            .checked_sub(self.balance)
            .ok_or(ParserError::OperationOverflows)?;

        Ok(fee)
    }
}

impl<'b> DisplayableItem for RegisterL1ValidatorTx<'b> {
    fn num_items(&self) -> Result<u8, ViewError> {
        // tx_info, node_id, weight, balance, fee, remaining_balance_owner, disable_owner
        let n_addresses =
            self.remaining_balance_owner.addresses.len() + self.disable_owner.addresses.len();
        Ok(5u8 + n_addresses as u8)
    }

    fn render_item(
        &self,
        item_n: u8,
        title: &mut [u8],
        message: &mut [u8],
        page: u8,
    ) -> Result<u8, ViewError> {
        use bolos::pic_str;
        use lexical_core::{write as itoa, Number};
        let mut buffer = [0; u64::FORMATTED_SIZE + 2];

        let n_remain_addresses = self.remaining_balance_owner.addresses.len();
        let n_disable_addresses = self.disable_owner.addresses.len();
        let prefix = pic_str!(b"0x"!);
        match item_n {
            0 => {
                let label = pic_str!(b"RegisterL1Val");
                title[..label.len()].copy_from_slice(label);
                let content = pic_str!(b"Transaction");
                handle_ui_message(content, message, page)
            }
            1 => self.node_id.render_item(0, title, message, page),
            2 => {
                let label = pic_str!(b"Weight");
                title[..label.len()].copy_from_slice(label);
                let buffer = itoa(self.weight, &mut buffer);
                handle_ui_message(buffer, message, page)
            }
            3 => {
                let label = pic_str!(b"Balance (AVAX)");
                title[..label.len()].copy_from_slice(label);

                let balance_buff = nano_avax_to_fp_str(self.balance, &mut buffer[..])
                    .map_err(|_| ViewError::Unknown)?;

                handle_ui_message(balance_buff, message, page)
            }
            4 => {
                let label = pic_str!(b"Fee(AVAX)");
                title[..label.len()].copy_from_slice(label);

                let fee = self.fee().map_err(|_| ViewError::Unknown)?;
                let fee_buff =
                    nano_avax_to_fp_str(fee, &mut buffer[..]).map_err(|_| ViewError::Unknown)?;

                handle_ui_message(fee_buff, message, page)
            }
            x if x >= 5 && x < (5 + n_remain_addresses as u8) => {
                let label = pic_str!(b"Remaining Owner");
                title[..label.len()].copy_from_slice(label);

                let mut out = [0; ADDRESS_LEN * 2 + 2];
                let mut sz = prefix.len();
                out[..prefix.len()].copy_from_slice(&prefix[..]);

                sz += hex_encode(
                    self.remaining_balance_owner.addresses[x as usize - 5],
                    &mut out[prefix.len()..],
                )
                .map_err(|_| ViewError::Unknown)?;

                handle_ui_message(&out[..sz], message, page)
            }
            x if x >= (5 + n_remain_addresses as u8)
                && x < (5 + n_remain_addresses as u8 + n_disable_addresses as u8) =>
            {
                let label = pic_str!(b"Disable Owner");
                title[..label.len()].copy_from_slice(label);

                let mut out = [0; ADDRESS_LEN * 2 + 2];
                let mut sz = prefix.len();
                out[..prefix.len()].copy_from_slice(&prefix[..]);

                sz += hex_encode(
                    self.disable_owner.addresses[x as usize - 5 - n_remain_addresses as usize],
                    &mut out[prefix.len()..],
                )
                .map_err(|_| ViewError::Unknown)?;

                handle_ui_message(&out[..sz], message, page)
            }
            _ => Err(ViewError::NoData),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::prelude::v1::*;

    use crate::parser::snapshots_common::ReducedPage;
    use zuit::Page;

    const DATA: &[u8] = &[];

    include!("testvectors/register_l1_validator.rs");
    #[test]
    fn parse_register_l1_validator() {
        let remaining_balance_owner_address = [
            0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e,
            0x7f, 0x80, 0x81, 0x82, 0x83, 0x84,
        ];

        let disable_owner_address = [
            0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f, 0x90, 0x91, 0x92, 0x93, 0x94, 0x95, 0x96,
            0x97, 0x98, 0x99, 0x9a, 0x9b, 0x9c,
        ];
        let (_, tx) = RegisterL1ValidatorTx::from_bytes(REGISTER_L1_VALIDATOR_DATA).unwrap();

        assert_eq!(
            tx.remaining_balance_owner.addresses[0],
            remaining_balance_owner_address
        );
        assert_eq!(tx.disable_owner.addresses[0], disable_owner_address);
        assert_eq!(tx.weight, 11357690822530343844);
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn ui_register_l1_validator() {
        for (i, data) in [REGISTER_L1_VALIDATOR_DATA].iter().enumerate() {
            println!("-------------------- Register L1 Validator TX #{i} ------------------------");
            let (_, tx) = RegisterL1ValidatorTx::from_bytes(data).unwrap();

            let items = tx.num_items().expect("Overflow?");

            let mut pages = Vec::<Page<18, 1024>>::with_capacity(items as usize);
            for i in 0..items {
                let mut page = Page::default();

                tx.render_item(i as _, &mut page.title, &mut page.message, 0)
                    .unwrap();

                pages.push(page);
            }

            let reduced = pages.iter().map(ReducedPage::from).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(reduced);
        }
    }
}
