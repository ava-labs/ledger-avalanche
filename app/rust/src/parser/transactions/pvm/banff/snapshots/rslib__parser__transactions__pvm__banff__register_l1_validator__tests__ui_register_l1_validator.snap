---
source: src/parser/transactions/pvm/banff/register_l1_validator.rs
assertion_line: 302
expression: reduced
---
[
    "RegisterL1Val": "Transaction",
    "Validator": "NodeID-KUhfajVUcCf2g5UXsZ1EXZpFYryppFRcZ",
    "Weight": "1",
    "Balance (AVAX)": "0.1",
    "Fee(AVAX)": "0.00004679",
    "Rem Addr": "0xdd9103b686299295f518f33ea3b6e767d06ead89",
    "Disabler": "0xdd9103b686299295f518f33ea3b6e767d06ead89",
]
